web/src/engine/osk/src/keyboard-layout/oskKey.ts

```diff
@@ -310,47 +294,73 @@ export default abstract class OSKKey {
-      // at least and doesn't combine with diacritic marks. For consistent display, it may be
-      // necessary to build a custom font that does not depend on renderer choices for base
-      // mark display -- e.g. create marks with custom base included, potentially even on PUA
-      // code points and use those in rendering the OSK. See #3039 for more details.
-      // keyText = '\u25cc' + keyText;
-
-      if(vkbd.isRTL) {
-        // Add the RTL marker to ensure it displays properly.
-        keyText = '\u200f' + keyText;
-      }
+    if(vkbd.isRTL) {
+      // Add the RTL marker to ensure it displays properly.
+      keyText = '\u200f' + keyText;
     }
 
-    ts.fontSize = this.getIdealFontSize(vkbd, keyText, styleSpec);
-
     // Finalize the key's text.
     t.innerText = keyText;
 
     return t;
   }
 
-  public refreshLayout(vkbd: VisualKeyboard) {
+  public resetFontPrecalc() {
+    this._fontFamily = undefined;
+    this._fontSize = undefined;
+    this.label.style.fontSize = '';
+  }
+
+  public refreshLayout(layoutParams: KeyLayoutParams) {
+    // Avoid doing any font-size related calculations if there's no text to display.
+    if(this.spec.sp == ButtonClasses.spacer || this.spec.sp == ButtonClasses.blank) {
+      return () => {};
+    }
+
+    // Attempt to detect static but key-specific style properties if they haven't yet
+    // been detected.
+    if(this._fontFamily === undefined) {
+      const lblStyle = getComputedStyle(this.label);
+
+      // Abort if the element is not currently in the DOM; we can't get any info this way.
+      if(!lblStyle.fontFamily) {
+        return () => {};
+      }
+      this._fontFamily = lblStyle.fontFamily;
+
+      // Detect any difference in base (em) font size and that which is computed for the key itself.
+      const computedFontSize = new ParsedLengthStyle(lblStyle.fontSize);
+      const layoutFontSize = layoutParams.layoutFontSize;
+      if(layoutFontSize.absolute) {
+        // rather than just straight-up taking .layoutFontSize
+        this._fontSize = computedFontSize;
+      } else {
+        const baseEmFontSize = layoutParams.baseEmFontSize;
+        const baseFontSize = layoutFontSize.scaledBy(baseEmFontSize.val);
+        const localFontScaling = computedFontSize.val / baseFontSize.val;
+        this._fontSize = ParsedLengthStyle.forScalar(localFontScaling);
+      }
+    }
+
     // space bar may not define the text span!
     if(this.label) {
       if(!this.label.classList.contains('kmw-spacebar-caption')) {
         // Do not use `this.keyText` - it holds *___* codes for special keys, not the actual glyph!
         const keyCapText = this.label.textContent;
-        this.label.style.fontSize = this.getIdealFontSize(vkbd, keyCapText, this.btn.style);
+        const fontSize = this.getIdealFontSize(keyCapText, layoutParams);
+        return () => {
+          this.label.style.fontSize = fontSize.styleString;
+        };
```

Comment:
This is kinda ugly and it's unclear why this is postponed and how it gets called

Reply:
This is the other side of that `keyTextClosure` you were asking about - its actual definition.

Admittedly, yeah, the `keyTextClosure` aspect is probably the least-clear thing in this PR.  It got tricky to forward well due to the inheritance structures involved.